VoodooI2CHID/VoodooI2CKeyboardHIDEventDriver.cpp

```diff
@@ -0,0 +1,479 @@
+        deviceClassMinor2 != kBluetoothDeviceClassMinorPeripheral2DigitizerTablet &&    // Magic Touchpad
+        deviceClassMinor2 != kBluetoothDeviceClassMinorPeripheral2DigitalPen)           // Wacom Tablet
+        return;
+    
+    attached_hid_pointer_devices->setObject(newService);
+    IOLog("%s: Bluetooth pointer HID device published: %s, # devices: %d\n", getName(), path, attached_hid_pointer_devices->getCount());
+}
+
+void VoodooI2CKeyboardHIDEventDriver::notificationHIDAttachedHandlerGated(IOService *newService, IONotifier *notifier) {
+    char path[MAX_PATH_LEN];
+    int len = MAX_PATH_LEN;
+    memset(path, 0, len);
+    newService->getPath(path, &len, gIOServicePlane);
+    
+    if (notifier == usb_hid_publish_notify) {
+        IORegistryEntry* hid_child = OSDynamicCast(IORegistryEntry, newService->getChildEntry(gIOServicePlane));
+        
+        if (!hid_child)
+            return;
+
+        OSNumber* primary_usage_page = OSDynamicCast(OSNumber, hid_child->getProperty(kIOHIDPrimaryUsagePageKey));
+        OSNumber* primary_usage = OSDynamicCast(OSNumber, hid_child->getProperty(kIOHIDPrimaryUsageKey));
+        
+        if (!primary_usage_page || !primary_usage)
+            return;
+        
+        // ignore touchscreens
+
+        if (primary_usage_page->unsigned8BitValue() != kHIDPage_Digitizer &&
+            primary_usage->unsigned8BitValue() != kHIDUsage_Dig_TouchScreen) {
+            
+            attached_hid_pointer_devices->setObject(newService);
+            IOLog("%s: USB pointer HID device published: %s, # devices: %d\n", getName(), path, attached_hid_pointer_devices->getCount());
+        }
+    }
+    
+    if (notifier == usb_hid_terminate_notify) {
+        attached_hid_pointer_devices->removeObject(newService);
+        IOLog("%s: USB pointer HID device terminated: %s, # devices: %d\n", getName(), path, attached_hid_pointer_devices->getCount());
+    }
+    
+    if (notifier == bluetooth_hid_publish_notify) {
+        bluetoothHIDAttached(newService, path);
+    }
+    
+    if (notifier == bluetooth_hid_terminate_notify) {
+        attached_hid_pointer_devices->removeObject(newService);
+        IOLog("%s: Bluetooth pointer HID device terminated: %s, # devices: %d\n", getName(), path, attached_hid_pointer_devices->getCount());
+    }
+}
+
+bool VoodooI2CKeyboardHIDEventDriver::notificationHIDAttachedHandler(void *refCon,
+                                                                     IOService *newService,
+                                                                     IONotifier *notifier) {
+    IOCommandGate::Action action = OSMemberFunctionCast(IOCommandGate::Action,
+                                                        this,
+                                                        &VoodooI2CKeyboardHIDEventDriver::notificationHIDAttachedHandlerGated);
+    
+    command_gate->runAction(action, newService, notifier);
+    return true;
```

Comment:
Returning whether `runAction` succeeded may be more appropriate, if it can actually fail anyway.